Src/AutoFixture/RegularExpressionGenerator.cs

```diff
@@ -28,16 +38,18 @@ public object Create(object request, ISpecimenContext context)
                 return new NoSpecimen();
             }
 
-            return GenerateRegularExpression(regularExpressionRequest);
+            return this.GenerateRegularExpression(regularExpressionRequest);
         }
 
-        private static object GenerateRegularExpression(RegularExpressionRequest request)
+        private object GenerateRegularExpression(RegularExpressionRequest request)
         {
             string pattern = request.Pattern;
 
             try
             {
-                string regex = new Xeger(pattern).Generate();
+                // Use the Xeger constructor overload that that takes an instance of Random.
+                // Otherwise identically strings can be generated, if regex are generated within short time.
+                string regex = new Xeger(pattern, this.random).Generate();
```

Comment:
As per my knowledge `Random` is not thread safe. See e.g. [here](https://stackoverflow.com/a/3049478/2009373). Previously we created a random instance per request (probably internally within Xeger), so we didn't have that problem.

Fixture by design can be used from multiple threads simultaneously, so we should take care of multi-threading.

Reply:
Ok, I'll look into it.
Should we use the code from https://codeblog.jonskeet.uk/2009/11/04/revisiting-randomness/ ? 

Basically doing `Xeger(pattern, ThreadLocalRandom.Instance)`

This way we could abstract threadsafety concerns out of RegularExpressionGenerator... 🤔, and maybe reuse it elsewhere.

Reply:
@MagnusMikkelsen you can just follow the pattern used in all the other generators, e.g.

https://github.com/AutoFixture/AutoFixture/blob/d3e14a818b933d948ea82464b197a5aeb152f070/Src/AutoFixture/RandomNumericSequenceGenerator.cs#L144

Reply:
@moodmosaic @MagnusMikkelsen If there is any way to do that without locks, please prefer that way. Introducing locking makes it hard to optimize performance later. When you have a bit suite, AutoFixture already takes substantial amount of time due to uncached reflection is uses everywhere - we shouldn't degrade it even more 😟 

In this particular case, it might be better to abstract out random generation and create generator, which uses e.g. `Interlocked`. Or, for instance, create thread-local random, but use interlocked to generate seed.

Reply:
Interlocked.CompareExchange or (IIRC) the Interlocked-extras by Jeffrey Richter in _CLR via C#_ can be candidates. 

- Unless you measure it, you can’t actually know how much ‘faster’ it’s going to be
- If speed matters, perhaps we can consider #928 — or [look into other PRNGs](https://github.com/AutoFixture/AutoFixture/issues/1040)

Though right now, I’d suggest we stick with `lock` as done in all other places and be consistent.

Reply:
Not sure how to use Interlocked.CompareExchange for this?

I did a couple of measurements, running the test below.
The times are measured using dotTrace, and are the total time and own time for `RegularExpressionGenerator.GenerateRegularExpression`.
The last test (ThreadLocal) took 12s. Seems to make sense, given the `GenerateRegularExpression` used some 85 seconds in total on 8 threads.

|					|Total Time [ms]  |	Own  [ms] |
|---------------------------|------------------|--------------|
|Current: shared `Random`|  83387               |  0                |
|Original: No injected `Random` | 88926 |	10 |
|New `Random` each time<br>locking on generate seed | 100843 |	0 |
| Sharing `ThreadLocal` <br> locking when generating `ThreadLocal` | 84848 |	27 |

(It is beside the point of this PR - but the `Xeger.ctor` seems to take most of this time - 81000ms total time. `Generate` is about 550ms total time. Might be an idea to cache `Xeger`? Fx `dictionary[pattern]=Xeger`)

``` cs
        [Fact]
        public void CreateManyAnonymousInParallelWithRegularExpressionValidatedTypeReturnsCorrectResult()
        {
            // Arrange
            var fixture = new Fixture();
            var result = Enumerable
                .Range(0, 10000)
                .AsParallel()
                .WithDegreeOfParallelism(8)
                .Select(_ => fixture.Create<RegularExpressionValidatedType>())
                .ToArray();

            // Assert

            Assert.All(result, r => Assert.Matches(RegularExpressionValidatedType.Pattern, r.Property));
            Assert.Equal(result.Distinct(), result);
        }
```

Reply:
> Interlocked-extras by Jeffrey Richter in _CLR via C#_ can be candidates.
Haven't got that book - so I don't it, do you have any links with examples? I'd like to learn new tricks 😁